plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,941 @@
+        # velocities
+        self.max_theta_hit_grid = self._max_theta_hit_grid()
+
+        # ************************************************************************
+        # Define the detector plane
+        # ************************************************************************
+
+        # Load or calculate the detector hdir
+        if detector_hdir is not None:
+            self.det_hdir = detector_hdir / np.linalg.norm(detector_hdir)
+        else:
+            self.det_hdir = self._default_detector_hdir()
+
+        # Calculate the detector vdir
+        ny = np.cross(self.det_hdir, self.det_n)
+        self.det_vdir = -ny / np.linalg.norm(ny)
+
+        # ************************************************************************
+        # Validate the E and B fields
+        # ************************************************************************
+
+        req_quantities = ["E_x", "E_y", "E_z", "B_x", "B_y", "B_z"]
+        for rq in req_quantities:
+
+            # Error check that grid contains E and B variables required
+            if rq not in self.grid.quantities:
+                warnings.warn(
+                    f"{rq} not specified for provided grid."
+                    "This quantity will be assumed to be zero.",
+                    RuntimeWarning,
+                )
+                # If missing, warn user and then replace with an array of zeros
+                unit = self.grid._recognized_quantities[rq].unit
+                arg = {rq: np.zeros(self.grid.shape) * unit}
+                self.grid.add_quantities(**arg)
+
+            # Check that there are no infinite values
+            if not np.isfinite(self.grid[rq]).all():
+                raise ValueError(
+                    f"Input arrays must be finite: {rq} contains "
+                    "either NaN or infinite values."
+                )
+
+            # Check that the max values on the edges of the arrays are
+            # small relative to the maximum values on that grid
+            arr = np.abs(self.grid[rq])
+            edge_max = np.max(
+                np.array(
+                    [
+                        np.max(arr[0, :, :]),
+                        np.max(arr[-1, :, :]),
+                        np.max(arr[:, 0, :]),
+                        np.max(arr[:, -1, :]),
+                        np.max(arr[:, :, 0]),
+                        np.max(arr[:, :, -1]),
+                    ]
+                )
+            )
+
+            if edge_max > 1e-3 * np.max(arr):
+                unit = grid.recognized_quantities[rq].unit
+                warnings.warn(
+                    "Fields should go to zero at edges of grid to avoid "
+                    f"non-physical effects, but a value of {edge_max:.2E} {unit} was "
+                    f"found on the edge of the {rq} array. Consider applying a "
+                    "envelope function to force the fields at the edge to go to "
+                    "zero.",
```

Comment:
I guess, if you have the ability to expand the grid to cover a larger area? I was looking at it from the perspective of having a grid of positions that is already specified.

Reply:
But the grid, as i understand it, is mostly a computational object onto which you project your simulated or experimental fields, so it would not in principle sound impossible to make the grid larger?

Reply:
Fair enough - if you have the option (eg. re-run a bigger simulation or create a larger analytical field) then that would work too. I've changed the text accordingly: 

```
"Fields should go to zero at edges of grid to avoid "
                    f"non-physical effects, but a value of {edge_max:.2E} {unit} was "
                    f"found on the edge of the {rq} array. Consider using a "
                    "larger grid (with lower fields at the edges), or applying an "
                    "envelope function to force the fields at the edge to go to "
                    "zero.",
```